### PR TITLE
Add compile-time support for single-precision Celeritas

### PR DESCRIPTION
**Pull request by sethrj**
This adds a new configuration option `CELERITAS_REAL_TYPE` which can be float or double and allows the whole Celeritas+ORANGE system to work with single-precision arithmetic. To make the PR more reviewable I'm splitting it into two, of which this is the first half:
- Add CMake option and constraints for when it can work
- Fix type errors (narrowing conversions of non-constant expressions) in src/ and app/
- Always use double precision for IO and Geant4 interaction
- Add tolerances/fudge factors specific to single precision
- Add an LLDB formatter to make debugging Spans super easy! (see celeritas/scripts/dev/celerlldb.py)
- Fix a propagation failure from having a zero-length chord (for steps that are far too small)
- Add assertions to check for nonzero movement (which happens a lot in single precision but would also break double-precision)

This adds a single-precision build (with no testing) to the CI. A follow-on PR will fix (most of) the tests.

### CODE REVIEW

File: src/celeritas/io/EventReader.cc [hunk 1]
```diff
@@ -115,21 +115,24 @@ auto EventReader::operator()() -> result_type
...
-        primary.position = {pos.x() * units::centimeter,
-                            pos.y() * units::centimeter,
-                            pos.z() * units::centimeter};
+        auto to_cm = [](double v) {
+            return static_cast<real_type>(v) * units::centimeter;
+        };
+        primary.position = {to_cm(pos.x()), to_cm(pos.y()), to_cm(pos.z())};
 
         // Get the lab-frame time [s]
         primary.time = pos.t() * units::centimeter / constants::c_light;
```
Comment (amandalund) [inline src/celeritas/io/EventReader.cc]: Should we do this for time as well?
```suggestion
        primary.time = to_cm(pos.t()) / constants::c_light;
```

**Comment by sethrj** [comment]
@pcanal @amandalund ready for review :)

File: CMakeLists.txt [hunk 2]
```diff
@@ -271,9 +282,19 @@ if(CELERITAS_USE_ROOT)
...
+    find_package(SWIG 4.1 REQUIRED)
+  endif()
+
+  if(CELERITAS_REAL_TYPE STREQUAL "float")
+    celeritas_error_incompatible_option(
+      "Single-precision runtime is not compatible with SWIG"
+      CELERITAS_USE_SWIG
+      OFF
+    )
+  endif()
```
Comment (pcanal) [inline CMakeLists.txt]:
```suggestion
  if(CELERITAS_REAL_TYPE STREQUAL "float")
    celeritas_error_incompatible_option(
      "Single-precision runtime is not compatible with SWIG"
      CELERITAS_USE_SWIG
      OFF
    )
  else()
    if(NOT SWIG_FOUND)
      # Note: 4.1 is required for C++17 parsing
      find_package(SWIG 4.1 REQUIRED)
    endif()
  endif()
```
No point in complaining twice if using single precision and not finding SWIG.

File: app/demo-interactor/KNDemoRunner.cc [hunk 3]
```diff
@@ -136,8 +136,9 @@ auto KNDemoRunner::operator()(KNDemoRunArgs args) -> result_type
     }
 
     // Copy integrated energy deposition
-    result.edep.resize(detector_params.tally_grid.size);
-    celeritas::app::finalize(params, state, make_span(result.edep));
+    std::vector<real_type> edep(detector_params.tally_grid.size);
```
Comment (pcanal) [inline app/demo-interactor/KNDemoRunner.cc:139]: That pattern may (or may not) be worth a wrapper.

File: src/accel/LocalTransporter.cc [hunk 4]
```diff
@@ -105,8 +105,8 @@ void LocalTransporter::Push(G4Track const& g4track)
 
     track.particle_id = particles_->find(
         PDGNumber{g4track.GetDefinition()->GetPDGEncoding()});
-    track.energy = units::MevEnergy{
-        convert_from_geant(g4track.GetKineticEnergy(), CLHEP::MeV)};
+    track.energy = units::MevEnergy(
```
Comment (pcanal) [inline src/accel/LocalTransporter.cc:108]: Why the switch from `{}` to `()` ?

File: src/orange/Constants.hh [hunk 5]
```diff
@@ -4,6 +4,7 @@
 // SPDX-License-Identifier: (Apache-2.0 OR MIT)
 //---------------------------------------------------------------------------//
 //! \file orange/Constants.hh
+//! \todo move to corecel/
```
Comment (pcanal) [inline src/orange/Constants.hh:7]: +1

**Review by pcanal — COMMENTED** [review]
See comments.

**Review by amandalund — APPROVED** [review]
Looks good to me!

File: CMakeLists.txt (hunk 2 above)
Comment (sethrj) [reply CMakeLists.txt]: Yeah, the reason I had it this way was uniformity with the other packages that have to be loaded first to determine compatibility. But are you manually requesting SWIG? Or is it just finding too old a version on your system?

File: src/accel/LocalTransporter.cc (hunk 4 above)
Comment (sethrj) [reply src/accel/LocalTransporter.cc:108]: C++11 brace initialization prohibits casting that could lose precision; so either I could manually insert `static_cast` (which would look like a static cast of double to double in most cases) or switch to parens. I'm not a huge fan of either option...

File: src/orange/Constants.hh (hunk 5 above)
Comment (sethrj) [reply src/orange/Constants.hh:7]: Maybe we need another issue like https://github.com/celeritas-project/celeritas/issues/491 — we should reorganize before 1.0 as well as renaming from #721

File: app/demo-interactor/KNDemoRunner.cc (hunk 3 above)
Comment (sethrj) [reply app/demo-interactor/KNDemoRunner.cc:139]: I'd rather delete this demo app 😅  in fact that may be a good "cleanup" activity for 1.0.

File: CMakeLists.txt (hunk 2 above)
Comment (pcanal) [reply CMakeLists.txt]: I dont have SWIG at all.

Comment (sethrj) [reply CMakeLists.txt]: Then it should have automatically set `CELERITAS_USE_SWIG=OFF` it it wasn't specified... do you have it on in your cmake user presets?

Comment (pcanal) [reply CMakeLists.txt]: No I did not request SWIG explicitly [independently, I don't use a preset].  Swig got turned on automatically when I switch my build from VecGeom to Orange ... but this is unrelated to the change and logic in this PR.

File: src/accel/LocalTransporter.cc (hunk 4 above)
Comment (pcanal) [reply src/accel/LocalTransporter.cc:108]: On the other hand the warning is real ... there is a narrowing ... Using paranthesis to mark the narrowing as been checked it is 'okay' is quite 'subtle'.

File: CMakeLists.txt (hunk 2 above)
Comment (sethrj) [reply CMakeLists.txt]: That shouldn't have happened. I can't think of why it would...

File: src/accel/LocalTransporter.cc (hunk 4 above)
Comment (sethrj) [reply src/accel/LocalTransporter.cc:108]: It's like the difference between explicitly and implicitly constructing an object... C++11 tried to make things better with brace initialization but I think it made things worse. (Also consider construction  for `struct Foo`: `Foo a;` doesn't initialize it, `Foo a{};` does 🤪 )

File: CMakeLists.txt (hunk 2 above)
Comment (pcanal) [reply CMakeLists.txt]: Me neither ... I'll mention it again if/when I see the issue.

File: src/accel/LocalTransporter.cc (hunk 4 above)
Comment (pcanal) [reply src/accel/LocalTransporter.cc:108]: and `Foo a();` declares a function :)

Comment (sethrj) [reply src/accel/LocalTransporter.cc:108]: Most vexing!

**Pull request merged**